src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/KeyValueStoreProcessor.cs

```diff
@@ -0,0 +1,281 @@
+
+    private List<KeyValueStoreAggregateHistory> _aggregatesToAdd = new();
+    private List<KeyValueStoreEntryHistory> _entriesToAdd = new();
+
+    public KeyValueStoreProcessor(ProcessorContext context)
+    {
+        _context = context;
+    }
+
+    public void VisitUpsert(CoreModel.Substate substateData, ReferencedEntity referencedEntity, long stateVersion)
+    {
+        if (substateData is CoreModel.GenericKeyValueStoreEntrySubstate genericKeyValueStoreEntry)
+        {
+            var kvStoreEntryLookup = new KeyValueStoreChangePointerLookup(
+                referencedEntity.DatabaseId,
+                stateVersion,
+                (ValueBytes)genericKeyValueStoreEntry.Key.KeyData.GetDataBytes()
+            );
+
+            _changePointers.GetOrAdd(kvStoreEntryLookup, l =>
+            {
+                _changeOrder.Add(kvStoreEntryLookup);
+
+                return new KeyValueStoreChangePointer(referencedEntity, genericKeyValueStoreEntry);
+            });
+        }
+    }
+
+    public void ProcessChanges()
+    {
+        foreach (var lookup in _changeOrder)
+        {
+            var change = _changePointers[lookup];
+
+            KeyValueStoreAggregateHistory aggregate;
+
+            if (!_mostRecentAggregates.TryGetValue(lookup.KeyValueStoreEntityId, out var previousAggregate) || previousAggregate.FromStateVersion != lookup.StateVersion)
+            {
+                aggregate = new KeyValueStoreAggregateHistory
+                {
+                    Id = _context.Sequences.KeyValueStoreAggregateHistorySequence++,
+                    FromStateVersion = lookup.StateVersion,
+                    KeyValueStoreEntityId = lookup.KeyValueStoreEntityId,
+                    KeyValueStoreEntryIds = new List<long>(),
+                };
+
+                if (previousAggregate != null)
+                {
+                    aggregate.KeyValueStoreEntryIds.AddRange(previousAggregate.KeyValueStoreEntryIds);
+                }
+
+                _aggregatesToAdd.Add(aggregate);
+                _mostRecentAggregates[lookup.KeyValueStoreEntityId] = aggregate;
+            }
+            else
+            {
+                aggregate = previousAggregate;
+            }
+
+            var entryLookup = new KeyValueStoreEntryDbLookup(lookup.KeyValueStoreEntityId, lookup.Key);
+
+            var isDeleted = change.KeyValueStoreEntry.Value == null;
+            var entry = new KeyValueStoreEntryHistory
+            {
+                Id = _context.Sequences.KeyValueStoreEntryHistorySequence++,
+                KeyValueStoreEntityId = lookup.KeyValueStoreEntityId,
+                FromStateVersion = lookup.StateVersion,
+                Key = change.KeyValueStoreEntry.Key.KeyData.GetDataBytes(),
+                IsLocked = change.KeyValueStoreEntry.IsLocked,
+                IsDeleted = isDeleted,
+                Value = isDeleted ? null : change.KeyValueStoreEntry.Value!.Data.StructData.GetDataBytes(),
```

Comment:
Absolutely minor notice: why do we need `!.` if we already know it is not null (`isDeleted == false`)?